### PR TITLE
MEN-4630: New chapter about how to use the device-side API

### PR DESCRIPTION
**Pull request by tranchitella**
Changelog: title

Signed-off-by: Fabio Tranchitella <fabio.tranchitella@northern.tech>

### CODE REVIEW

**Comment by lluiscampos** [comment]
This PR should go into the `feature-update_control_maps` branch, right? I don't have the overview of what we have added there, but we need to make sure that what we are writing in this PR builds on top of what is there already

**Comment by tranchitella** [comment]
@lluiscampos right, sorry

File: 03.Client-installation/06.Use-the-device-side-api/docs.md [hunk 1]
```diff
@@ -0,0 +1,147 @@
...
+marked as failed by updating the Update Control Map of the Mender client.
+
+You can manage the update progress using the [Device-side APIs](../../201.Device-side-API),
+updating the Mender client's Update Control Map using the *SetUpdateControlMap* method of the
+*Mender Update Management API* .
+
+Each Update Control Map is identified by a unique identifier (UUID) and can optionally
+specify a priority. Once set, the application will receive as return value an expiration
+time out expressed in seconds. It's the application's responsibility to refresh the
+Update Control Map before the expiration time is over to ensure the Update Control Map
```
Comment (oleorhagen) [inline 03.Client-installation/06.Use-the-device-side-api/docs.md:41]: Suggestion:
I'm thinking `expiration time` is enough, and we can drop the `out`

File: 03.Client-installation/06.Use-the-device-side-api/docs.md [hunk 2]
```diff
@@ -0,0 +1,147 @@
...
+the priority (higher priority wins). The Mender Client then evaluates the resulting Update
+Control Map.
+
+## Example applications
+
+A set of example applications consuming the Device-side API can be found in the
+[mender-client-python-example](https://github.com/mendersoftware/mender-client-python-example)
+repository.
+
+The examples require Python 3, and its package manager `pip`. We also use `git`
```
Comment (oleorhagen) [inline 03.Client-installation/06.Use-the-device-side-api/docs.md:54]: Opinion:
I think the installation instructions should be kept in the `example` directory only

**Review by oleorhagen — APPROVED** [review]

File: 03.Client-installation/06.Use-the-device-side-api/docs.md [hunk 3]
```diff
@@ -0,0 +1,147 @@
+---
+title: Use the Device-side API
+taxonomy:
+    category: tutorial
+---
+
+The Device-side API constitutes the only public API of the Mender client. It is a thin layer
```
Comment (lluiscampos) [inline 03.Client-installation/06.Use-the-device-side-api/docs.md:7]: Nitpick, but "API" is by default public. Suggestion:
```suggestion
The Device-side API constitutes the only public programmatic interface of the Mender client. It is a thin layer
```

File: 03.Client-installation/06.Use-the-device-side-api/docs.md [hunk 4]
```diff
@@ -0,0 +1,147 @@
...
+you can set **pause**.
+
+Once *paused*, the update process can either be *resumed* or *canceled* and the whole deployment
+marked as failed by updating the Update Control Map of the Mender client.
+
+You can manage the update progress using the [Device-side APIs](../../201.Device-side-API),
+updating the Mender client's Update Control Map using the *SetUpdateControlMap* method of the
+*Mender Update Management API* .
+
+Each Update Control Map is identified by a unique identifier (UUID) and can optionally
```
Comment (lluiscampos) [inline 03.Client-installation/06.Use-the-device-side-api/docs.md:38]: Being a tutorial, I would add a link to Wikipedia for the UUID: https://en.wikipedia.org/wiki/Universally_unique_identifier

File: 03.Client-installation/06.Use-the-device-side-api/docs.md (hunk 2 above)
Comment (lluiscampos) [reply 03.Client-installation/06.Use-the-device-side-api/docs.md:54]: I agree. There is already a good deal of repetition, but at least for the installation let's keep it only in the other repo.

File: 03.Client-installation/06.Use-the-device-side-api/docs.md [hunk 5]
```diff
@@ -0,0 +1,147 @@
...
+
+## Example applications
+
+A set of example applications consuming the Device-side API can be found in the
+[mender-client-python-example](https://github.com/mendersoftware/mender-client-python-example)
+repository.
+
+The examples require Python 3, and its package manager `pip`. We also use `git`
+to get the source. For Debian based distributions, install the required dependencies
+with:
```
Comment (lluiscampos) [inline 03.Client-installation/06.Use-the-device-side-api/docs.md:56]: No dependencies being installed here (not explicitly, at least).

File: 03.Client-installation/06.Use-the-device-side-api/docs.md [hunk 6]
```diff
@@ -0,0 +1,147 @@
...
+When deploying an update to the device, the Mender Client will pause the processing
+of the deployment waiting for input from the user to proceed:
+
+![deployment paused](deployment-paused.jpg)
+
+Pressing `1`, `2` or `3`, the user can instruct the Mender Client to proceed with
+the installation of the software update, the reboot (if required by the artifact)
+and the commit of the update.
+
+For example, to continue with the installation of the artifact, you can press
```
Comment (lluiscampos) [inline 03.Client-installation/06.Use-the-device-side-api/docs.md:114]:
```suggestion
For example, to continue with the installation of the Artifact, you can press
```

File: 03.Client-installation/06.Use-the-device-side-api/docs.md [hunk 7]
```diff
@@ -0,0 +1,147 @@
...
+```bash
+git clone https://github.com/mendersoftware/mender-client-python-example mender-client-python-example
+sudo pip3 install mender-client-python-example
+```
+### Update flow control
+
+<!--AUTOVERSION: "mender-client-python-example/tree/%/examples/update_flow_control"/ignore-->
+The [Update flow control](https://github.com/mendersoftware/mender-client-python-example/tree/master/examples/update_flow_control)
+example application controls the update flow setting Update Control Maps in the Mender
+Client based on the user's input.
```
Comment (lluiscampos) [inline 03.Client-installation/06.Use-the-device-side-api/docs.md:67]:
```suggestion
client based on the user's input.
```

File: 06.Artifact-creation/04.State-scripts/docs.md [hunk 8]
```diff
@@ -125,17 +123,6 @@ In this case, application data like a user profile is stored in an SQLite databa
...
-
-Make sure to adjust [StateScriptRetryTimeoutSeconds](../../03.Client-installation/06.Configuration-file/50.Configuration-options/docs.md#statescriptretrytimeoutseconds), to enable this use case.
-
-![End user update confirmation state scripts](mender-state-machine-user-confirmation.png)
-
-!! Maximum wait time between `Sync` and `Download` state is 24 hours, after this period the update will be marked as failed by the Mender client. This happens because the Mender Artifact download link is generated in `Sync` state and it has a expiration time (24 hours).
-
```
Comment (lluiscampos) [inline 06.Artifact-creation/04.State-scripts/docs.md:138]: :+1:

**Review by lluiscampos — APPROVED** [review]
Looks good otherwise

File: 06.Artifact-creation/04.State-scripts/docs.md [hunk 9]
```diff
@@ -125,17 +123,6 @@ In this case, application data like a user profile is stored in an SQLite databa
 ![Application data migration state scripts](mender-state-machine-data-migration.png)
 
 
-#### Update confirmation by end user
```
Comment (kacf) [inline 06.Artifact-creation/04.State-scripts/docs.md:128]: I think it is premature to remove this section. See this comment that I made in the [blog post draft](https://docs.google.com/document/d/16eUZFe4vyc8ndvOKWlciT5kj2W8J2etnF8bOjJ-Ywy8/edit#):

> Actually, this won't work, the feature is "incomplete" without implementing https://tracker.mender.io/browse/MEN-4194 as well. Although the API allows for the correct sequence of actions to control the update, there is no way to notify the external process (device UI) that an update is available. You could hack around it by calling deployments/next yourself with the JWT token, but this is error prone and can give undesirable effects if called incorrectly. Certainly not something I would advertise.
> 
> With the currently available features, what we can more confidently support is an "on/off" switch, where the device can pause all updates during certain times (when doing something critical), and have them continue afterwards. However, the process won't know if any updates were actually in progress.

Comment (tranchitella) [reply 06.Artifact-creation/04.State-scripts/docs.md:128]: @kacf do you want me to reintroduce it, then?

Comment (kacf) [reply 06.Artifact-creation/04.State-scripts/docs.md:128]: Yes, could you? 😍

Comment (tranchitella) [reply 06.Artifact-creation/04.State-scripts/docs.md:128]: https://github.com/mendersoftware/mender-docs/pull/1491

**Pull request merged**